aten/src/ATen/native/mps/operations/BinaryOps.mm

```diff
@@ -21,19 +22,55 @@
 typedef MPSGraphTensor* (^BinaryOpBlock)(MPSGraph*, MPSGraphTensor*, MPSGraphTensor*);
 #define BinaryOpFn() MPSGraphTensor* (MPSGraph* mpsGraph, MPSGraphTensor* primary, MPSGraphTensor* secondary)
 
-void binaryOpTensor(const Tensor& self, const Tensor& other, const Tensor& output, std::string op_name, BinaryOpBlock binaryBlock)
+// Assign a precedence to types for comparing in boolean operation
+int type_precedence_number(const MPSDataType tensor_dtype) {
+
+  switch (tensor_dtype) {
+    case MPSDataTypeFloat32:
+      return 7;
+    case MPSDataTypeFloat16:
+      return 6;
+    case MPSDataTypeInt64:
+      return 5;
+    case MPSDataTypeInt32:
+      return 4;
+    case MPSDataTypeInt16:
+      return 3;
+    case MPSDataTypeInt8:
+      return 2;
+    case MPSDataTypeBool:
+      return 1;
+    default:
+      TORCH_CHECK_TYPE(false, "Unsupported data type")
+  }
+
+}
+
+// is_arithmetic lets us check for arithmetic ops, so we can cast the input to output type if needed
+// because in arithmetic ops the input and output types match
+void binaryOpTensor(const Tensor& self_t, const Tensor& other_t, const Tensor& output, std::string op_name, bool is_arithmetic, BinaryOpBlock binaryBlock)
 {
   // it's possible to receive empty tensors here
-  if (self.numel() == 0 || other.numel() == 0) {
+  if (self_t.numel() == 0 || other_t.numel() == 0) {
     return;
   }
   MPSStream* mpsStream = getCurrentMPSStream();
 
-  const bool is_self_scalar = self.dim() == 0;
-  const bool is_other_scalar = other.dim() == 0;
+  const bool is_self_scalar = self_t.dim() == 0;
+  const bool is_other_scalar = other_t.dim() == 0;
+
+  Tensor self = is_self_scalar ? self_t : self_t.contiguous(at::MemoryFormat::Contiguous);
```

Comment:
Why are we putting contiguous back here ?

Reply:
Removed it